cpu/atmega2560/periph/gpio.c

```diff
@@ -28,22 +27,169 @@
 
 #if GPIO_NUMOF
 
+/*
 typedef struct {
     void (*cb)(void);
 } gpio_state_t;
+*/
 
-/*static gpio_state_t config[GPIO_NUMOF]; */
-
+/* static gpio_state_t config[GPIO_NUMOF]; */
 
+/* static pin mapping */
+static const uint8_t gpio_pin_map[GPIO_NUMOF] = {
+#if GPIO_0_EN
+    [GPIO_0] = GPIO_0_PIN,
+#endif
+#if GPIO_1_EN
+    [GPIO_1] = GPIO_1_PIN,
+#endif
+#if GPIO_2_EN
+    [GPIO_2] = GPIO_2_PIN,
+#endif
+#if GPIO_3_EN
+    [GPIO_3] = GPIO_3_PIN,
+#endif
+#if GPIO_4_EN
+    [GPIO_4] = GPIO_4_PIN,
+#endif
+#if GPIO_5_EN
+    [GPIO_5] = GPIO_5_PIN,
+#endif
+#if GPIO_6_EN
+    [GPIO_6] = GPIO_6_PIN,
+#endif
+};
 
 int gpio_init_out(gpio_t dev, gpio_pp_t pushpull)
 {
-    return -1;
+    uint8_t pin;
+
+    pin = gpio_pin_map[dev];
+
+    switch (dev)
+    {
+#if GPIO_0_EN
+        case GPIO_0:
+            GPIO_0_DDR_REG |= (1 << pin);
+            break;
+#endif /* GPIO_0_EN */
+
+#if GPIO_1_EN
+        case GPIO_1:
+            GPIO_1_DDR_REG |= (1 << pin);
+            break;
+#endif /* GPIO_1_EN */
+
+#if GPIO_2_EN
+        case GPIO_2:
+            GPIO_2_DDR_REG |= (1 << pin);
+            break;
+#endif /* GPIO_2_EN */
+
+#if GPIO_3_EN
+        case GPIO_3:
+            GPIO_3_DDR_REG |= (1 << pin);
+            break;
+#endif /* GPIO_2_EN */
+
+#if GPIO_4_EN
+        case GPIO_4:
+            GPIO_4_DDR_REG |= (1 << pin);
+            break;
+#endif /* GPIO_4_EN */
+
+#if GPIO_5_EN
+        case GPIO_5:
+            GPIO_5_DDR_REG |= (1 << pin);
+            break;
+#endif /* GPIO_5_EN */
+
+#if GPIO_6_EN
+        case GPIO_6:
+            GPIO_6_DDR_REG |= (1 << pin);
+            break;
+#endif /* GPIO_6_EN */
+    };
+
+    switch(pushpull)
+    {
+        case GPIO_NOPULL:
+            gpio_clear(dev);
+            break;
+        case GPIO_PULLUP:
+            gpio_set(dev);
+            break;
+        case GPIO_PULLDOWN:
+            return -1;
+    };
```

Comment:
Did you overread my comment: "*I thought about it a bit. Let's just leave the pull-up configuration out for the first. It seems to be handled by hardware in a "senseful" way. Just add a comment for this here. Sorry for the confusion*" ?
As we read in the manual the pull-up handling is half automatic and not separately controllable for each pin. So let's just take it out.